File: CPH/permutation.rb
def permutation(arr, visited = {}, cur_per = [], ans = [])
  if cur_per.size == arr.size
    ans << cur_per[0..-1]
  else
    arr.each do |num|
      next if visited[num]

      visited[num] = true
      cur_per.push num
      permutation(arr, visited, cur_per, ans)

      visited[num] = false
      cur_per.pop
    end
  end

  ans
end

# check Python implementation which seems easier

puts "-----permutation of 1,2,3 is #{permutation([1, 2, 3])}"

File: CPH/permutation.py
def permutation(arr, start = 0, ans = [] ):
  if start >= len(arr):
      ans.append(arr[:])
      return

  for i in range(start, len(arr)):
    arr[i], arr[start] = arr[start], arr[i]
    permutation(arr, start+1, ans)
    arr[i], arr[start] = arr[start], arr[i]
  return ans
print(permutation([1,2,3]))